aiographite/aiographitesend.py

```diff
@@ -0,0 +1,449 @@
+
+
+
+	def generate_message_for_plaintext(self, listOfPlaintext):
+		"""
+			return the required message formate for protocol 'plaintext'
+			@param: 
+				listOfPlaintext: ["metric1 value1 timestamp1", "metric2 value2 timestamp2", ...]
+				type: List of String
+		"""
+		return "".join(listOfPlaintext)
+
+
+
+	def generate_message_for_pickle(self, listOfMetricTuples):
+		"""
+			@param: 
+				listOfMetricTuples: [(metric1, (timestamp1, value1), (metric2, (timestamp2, value2), ...]
+		"""
+		payload = pickle.dumps(listOfMetricTuples, protocol=2)
+		header = struct.pack("!L", len(payload))
+		message = header + payload
+		return message
+
+
+	def plaintext_protocol_formatted_data(self, metric, value, timestamp):
+		"""
+			@return: required data formate when sending data through 'plaintext' protocol
+			@return_type: String
+		"""
+		formatted_data = " ".join([metric, str(value), str(timestamp)])
+		return formatted_data + "\n"
+
+
+
+	def pickle_protocol_formatted_data(self, metric, value, timestamp):
+		"""
+			@return: required data formate when sending data through 'pickle' protocol
+			@return_type: Tuple
+		"""		
+		return (metric, (timestamp, value))
+
+
+
+	def send_message(self, message):
+		"""
+			@message: data ready to sent to graphite server
+		"""
+		total_sent = 0
+		message_size = len(message)
+		while total_sent < message_size:
+			try:
+				sent = self.socket.send(message[total_sent:].encode('ascii'))
+				if sent == 0:
+					raise RuntimeError("socket connection broken")
+				total_sent = total_sent + sent
+			except socket.gaierror as e:
+				raise AioGraphiteSendException("Fail to send data to %s, Error: %s" % (self.graphite_server_address, e)) 
+			except Exception as e:
+				raise e
+		return total_sent
+
+
+
+	def disconnect(self):
+		"""
+			Close the TCP connection 
+		"""
+		try:
+			self.socket.shutdown(1)
+		except AttributeError:
+			self.socket = None
+		except Exception:
+			self.socket = None
+		finally:
+			self.socket = None
+
+
+	def close_event_loop(self):
+		"""
+			Close Event Loop. 
+			No call should be made after event loop closed
+		"""
+		self.loop.close()
+
+
+
+	def to_graphite_valid_metric_name(self, metric_dir_list):
+		"""
+			@purpose:
+				Make metric name valid for graphite in case that the metric name includes 
+				any special character which is not supported by Graphite
+			@example: 
+				Assuming that 
+
+					Expected_Metric_Name  =  metaccounts.authentication.password.attempted
+
+				Then input metric_dir_list should be
+
+					metric_dir_list = [metaccounts, authentication, password, attempted]
+
+			@metric_dir_list: List of String
+		"""
+		return ".".join([metrics_name_to_graphite(dir_name) for dir_name in metric_dir_list])
+
+
+
+#########################################################
+#########################################################
+#########################################################
+
+
+# Module Instance Variable
+aiographite_send_instance = None
```

Comment:
I don't think we need those methods.

If someone wants to make a global aiographite_send, they can do that themselves.

In general, these types of globals are dangerous, because they can lead to race conditions and assumptions about configuration that wasn't correct.